File: library/analytics/src/main/java/org/cru/godtools/analytics/AdobeAnalyticsService.java
package org.cru.godtools.analytics;

import android.annotation.SuppressLint;
import android.app.Activity;
import android.app.Application;
import android.content.Context;
import android.net.Uri;
import android.os.Bundle;

import com.adobe.mobile.Analytics;
import com.adobe.mobile.Config;
import com.adobe.mobile.Visitor;

import org.cru.godtools.analytics.model.AnalyticsActionEvent;
import org.cru.godtools.analytics.model.AnalyticsBaseEvent;
import org.cru.godtools.analytics.model.AnalyticsScreenEvent;
import org.cru.godtools.analytics.model.AnalyticsSystem;
import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.lang.ref.WeakReference;
import java.util.Collections;
import java.util.HashMap;
import java.util.Locale;
import java.util.Map;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

import androidx.annotation.AnyThread;
import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.annotation.WorkerThread;
import me.thekey.android.TheKey;

import static me.thekey.android.Attributes.ATTR_GR_MASTER_PERSON_ID;
import static org.ccci.gto.android.common.compat.util.LocaleCompat.toLanguageTag;

public final class AdobeAnalyticsService implements AnalyticsService, Application.ActivityLifecycleCallbacks {
    /* Property Keys */
    private static final String KEY_APP_NAME = "cru.appname";
    private static final String KEY_MARKETING_CLOUD_ID = "cru.mcid";
    private static final String KEY_SSO_GUID = "cru.ssoguid";
    private static final String KEY_GR_MASTER_PERSON_ID = "cru.grmpid";
    private static final String KEY_LOGGED_IN_STATUS = "cru.loggedinstatus";
    private static final String KEY_SCREEN_NAME = "cru.screenname";
    private static final String KEY_SCREEN_NAME_PREVIOUS = "cru.previousscreenname";
    private static final String KEY_CONTENT_LANGUAGE = "cru.contentlanguage";
    public static final String KEY_CONTENT_LANGUAGE_SECONDARY = "cru.contentlanguagesecondary";
    private static final String KEY_EXIT_LINK = "cru.mobileexitlink";
    private static final String KEY_SHARE_CONTENT = "cru.shareiconengaged";
    private static final String KEY_SITE_SECTION = "cru.sitesection";
    private static final String KEY_SITE_SUB_SECTION = "cru.sitesubsection";

    /* Value constants */
    private static final String VALUE_GODTOOLS = "GodTools";
    private static final String VALUE_LOGGED_IN = "logged in";
    private static final String VALUE_NOT_LOGGED_IN = "not logged in";

    private final Context mContext;
    private final TheKey mTheKey;
    /**
     * Single thread executor to serialize events on a background thread.
     */
    private final Executor mAnalyticsExecutor = Executors.newSingleThreadExecutor();

    @NonNull
    private WeakReference<Activity> mActiveActivity = new WeakReference<>(null);
    private String mPreviousScreenName;

    private AdobeAnalyticsService(@NonNull final Context context) {
        mContext = context;
        mTheKey = TheKey.getInstance(mContext);
        Config.setContext(context);

        EventBus.getDefault().register(this);
        ((Application) mContext.getApplicationContext()).registerActivityLifecycleCallbacks(this);
    }

    @Nullable
    @SuppressLint("StaticFieldLeak")
    private static AdobeAnalyticsService sInstance;

    @NonNull
    public static synchronized AdobeAnalyticsService getInstance(@NonNull final Context context) {
        if (sInstance == null) {
            sInstance = new AdobeAnalyticsService(context.getApplicationContext());
        }

        return sInstance;
    }

    // region Tracking Events

    @WorkerThread
    @Subscribe(threadMode = ThreadMode.ASYNC)
    public void onAnalyticsActionEvent(@NonNull final AnalyticsActionEvent event) {
        if (event.isForSystem(AnalyticsSystem.ADOBE)) {
            trackAction(event.getAction(), event, event.getAdobeAttributes());
        }
    }

    @WorkerThread
    @Subscribe(threadMode = ThreadMode.ASYNC)
    public void onAnalyticsScreenEvent(@NonNull final AnalyticsScreenEvent event) {
        if (event.isForSystem(AnalyticsSystem.ADOBE)) {
            trackState(event);
        }
    }

    @Override
    public void onTrackShareAction() {
        trackAction(ACTION_SHARE, null, Collections.singletonMap(KEY_SHARE_CONTENT, 1));
    }

    @Override
    public void onTrackExitUrl(@NonNull final Uri url) {
        trackAction(ACTION_EXIT_LINK, null, Collections.singletonMap(KEY_EXIT_LINK, url.toString()));
    }

    // region ActivityLifecycleCallbacks

    @Override
    public void onActivityCreated(final Activity activity, final Bundle savedInstanceState) {}

    @Override
    public void onActivityStarted(final Activity activity) {}

    @Override
    @MainThread
    public void onActivityResumed(final Activity activity) {
        final String guid = mTheKey.getDefaultSessionGuid();
        mActiveActivity = new WeakReference<>(activity);
        mAnalyticsExecutor.execute(() -> Config.collectLifecycleData(activity, baseContextData(guid, null)));
    }

    @Override
    public void onActivityPaused(final Activity activity) {
        if (mActiveActivity.get() == activity) {
            mActiveActivity = new WeakReference<>(null);
            mAnalyticsExecutor.execute(Config::pauseCollectingLifecycleData);
        }
    }

    @Override
    public void onActivityStopped(final Activity activity) {}

    @Override
    public void onActivitySaveInstanceState(final Activity activity, final Bundle outState) {}

    @Override
    public void onActivityDestroyed(final Activity activity) {}

    // endregion ActivityLifecycleCallbacks

    // endregion Tracking Events

    @AnyThread
    private void trackAction(@NonNull final String action, @Nullable final AnalyticsActionEvent event,
                             @Nullable final Map<String, ?> attributes) {
        final String guid = mTheKey.getDefaultSessionGuid();
        mAnalyticsExecutor.execute(() -> {
            final Map<String, Object> data = baseContextData(guid, event);
            if (mPreviousScreenName != null) {
                data.put(KEY_SCREEN_NAME, mPreviousScreenName);
            }
            if (attributes != null) {
                data.putAll(attributes);
            }
            Analytics.trackAction(action, data);
        });
    }

    @AnyThread
    private void trackState(@NonNull final AnalyticsScreenEvent event) {
        final String guid = mTheKey.getDefaultSessionGuid();
        mAnalyticsExecutor.execute(() -> {
            Analytics.trackState(event.getScreen(), stateContextData(guid, event));
            mPreviousScreenName = event.getScreen();
        });
    }

    /**
     * Visitor.getMarketingCloudId() may be blocking. So, we need to call it on a worker thread.
     */
    @WorkerThread
    private Map<String, Object> baseContextData(@Nullable final String guid, @Nullable final AnalyticsBaseEvent event) {
        final Map<String, Object> data = new HashMap<>();
        data.put(KEY_APP_NAME, VALUE_GODTOOLS);
        data.put(KEY_MARKETING_CLOUD_ID, Visitor.getMarketingCloudId());

        // login state
        data.put(KEY_LOGGED_IN_STATUS, guid != null ? VALUE_LOGGED_IN : VALUE_NOT_LOGGED_IN);
        if (guid != null) {
            data.put(KEY_SSO_GUID, guid);
            final String grMasterPersonId = mTheKey.getAttributes(guid).getAttribute(ATTR_GR_MASTER_PERSON_ID);
            if (grMasterPersonId != null) {
                data.put(KEY_GR_MASTER_PERSON_ID, grMasterPersonId);
            }
        }

        if (event != null) {
            final Locale contentLocale = event.getLocale();
            if (contentLocale != null) {
                data.put(KEY_CONTENT_LANGUAGE, toLanguageTag(contentLocale));
            }

            final String siteSection = event.getAdobeSiteSection();
            if (siteSection != null) {
                data.put(KEY_SITE_SECTION, siteSection);
            }
            final String siteSubSection = event.getAdobeSiteSubSection();
            if (siteSubSection != null) {
                data.put(KEY_SITE_SUB_SECTION, siteSubSection);
            }
        }

        return data;
    }

    /**
     * This method created the Data to be passed to Adobe my converting it to a HashMap
     *
     * @param guid  user id from key
     * @param event The analytics event we are generating context data for
     * @return The data to be tracked by Adobe Analytics
     */
    @WorkerThread
    private Map<String, Object> stateContextData(@Nullable final String guid,
                                                 @NonNull final AnalyticsScreenEvent event) {
        final Map<String, Object> data = baseContextData(guid, event);
        data.put(KEY_SCREEN_NAME_PREVIOUS, mPreviousScreenName);
        data.put(KEY_SCREEN_NAME, event.getScreen());
        return data;
    }
}

File: library/analytics/src/main/java/org/cru/godtools/analytics/adobe/AdobeAnalyticsService.kt
package org.cru.godtools.analytics.adobe

import android.app.Activity
import android.app.Application
import android.app.Application.ActivityLifecycleCallbacks
import android.os.Bundle
import androidx.annotation.AnyThread
import androidx.annotation.MainThread
import androidx.annotation.WorkerThread
import com.adobe.mobile.Analytics
import com.adobe.mobile.Config
import com.adobe.mobile.Visitor
import com.adobe.mobile.VisitorID.VisitorIDAuthenticationState
import com.karumi.weak.weak
import com.okta.oidc.net.response.UserInfo
import java.util.concurrent.Executor
import java.util.concurrent.Executors
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.SharingStarted.Companion.Eagerly
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.flow.stateIn
import org.ccci.gto.android.common.compat.util.LocaleCompat
import org.ccci.gto.android.common.okta.oidc.OktaUserProfileProvider
import org.ccci.gto.android.common.okta.oidc.net.response.grMasterPersonId
import org.ccci.gto.android.common.okta.oidc.net.response.ssoGuid
import org.cru.godtools.analytics.BuildConfig
import org.cru.godtools.analytics.model.AnalyticsActionEvent
import org.cru.godtools.analytics.model.AnalyticsBaseEvent
import org.cru.godtools.analytics.model.AnalyticsScreenEvent
import org.cru.godtools.analytics.model.AnalyticsSystem
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

/* Property Keys */
private const val ADOBE_ATTR_APP_NAME = "cru.appname"
private const val ADOBE_ATTR_MARKETING_CLOUD_ID = "cru.mcid"
private const val ADOBE_ATTR_SSO_GUID = "cru.ssoguid"
private const val ADOBE_ATTR_GR_MASTER_PERSON_ID = "cru.grmpid"
private const val ADOBE_ATTR_LOGGED_IN_STATUS = "cru.loggedinstatus"
private const val ADOBE_ATTR_SCREEN_NAME = "cru.screenname"
private const val ADOBE_ATTR_SCREEN_NAME_PREVIOUS = "cru.previousscreenname"
private const val ADOBE_ATTR_CONTENT_LANGUAGE = "cru.contentlanguage"
const val ADOBE_ATTR_LANGUAGE_SECONDARY = "cru.contentlanguagesecondary"
private const val ADOBE_ATTR_SITE_SECTION = "cru.sitesection"
private const val ADOBE_ATTR_SITE_SUB_SECTION = "cru.sitesubsection"

/* Value constants */
private const val VALUE_GODTOOLS = "GodTools"
private const val VALUE_LOGGED_IN = "logged in"
private const val VALUE_NOT_LOGGED_IN = "not logged in"

private const val VISITOR_ID_GUID = "ssoguid"
private const val VISITOR_ID_MASTER_PERSON_ID = "grmpid"
private const val VISITOR_ID_ECID = "ecid"

@Singleton
@OptIn(ExperimentalStdlibApi::class)
class AdobeAnalyticsService @Inject internal constructor(
    app: Application,
    eventBus: EventBus,
    oktaUserProfileProvider: OktaUserProfileProvider
) : ActivityLifecycleCallbacks {
    private val coroutineScope = CoroutineScope(Dispatchers.Default)

    private val analyticsExecutor: Executor = Executors.newSingleThreadExecutor()

    init {
        Config.setDebugLogging(BuildConfig.DEBUG)
        Config.setContext(app)
        eventBus.register(this)
        app.registerActivityLifecycleCallbacks(this)
    }

    // region Tracking Events
    @WorkerThread
    @Subscribe(threadMode = ThreadMode.ASYNC)
    fun onAnalyticsActionEvent(event: AnalyticsActionEvent) {
        if (event.isForSystem(AnalyticsSystem.ADOBE)) event.track()
    }

    @WorkerThread
    @Subscribe(threadMode = ThreadMode.ASYNC)
    fun onAnalyticsScreenEvent(event: AnalyticsScreenEvent) {
        if (event.isForSystem(AnalyticsSystem.ADOBE)) event.track()
    }

    // region ActivityLifecycleCallbacks
    private var activeActivity: Activity? by weak()

    override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) = Unit
    override fun onActivityStarted(activity: Activity) = Unit

    @MainThread
    override fun onActivityResumed(activity: Activity) {
        activeActivity = activity

        val userProfile = userProfileStateFlow.value
        analyticsExecutor.execute {
            Config.collectLifecycleData(activity, buildMap { baseContextData(userProfile) })
        }
    }

    @MainThread
    override fun onActivityPaused(activity: Activity) {
        if (activeActivity === activity) {
            activeActivity = null
            analyticsExecutor.execute { Config.pauseCollectingLifecycleData() }
        }
    }

    override fun onActivityStopped(activity: Activity) = Unit
    override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) = Unit
    override fun onActivityDestroyed(activity: Activity) = Unit
    // endregion ActivityLifecycleCallbacks
    // endregion Tracking Events

    private var previousScreenName: String? = null

    @AnyThread
    private fun AnalyticsActionEvent.track() {
        val userProfile = userProfileStateFlow.value
        analyticsExecutor.execute {
            Analytics.trackAction(action, buildMap {
                baseContextData(userProfile, this@track)
                previousScreenName?.let { put(ADOBE_ATTR_SCREEN_NAME, it) }
                adobeAttributes?.let { putAll(it) }
            })
        }
    }

    @AnyThread
    private fun AnalyticsScreenEvent.track() {
        val userProfile = userProfileStateFlow.value
        analyticsExecutor.execute {
            Analytics.trackState(screen, buildMap { stateContextData(userProfile, this@track) })
            previousScreenName = screen
        }
    }

    @WorkerThread
    private fun MutableMap<String, Any?>.baseContextData(userProfile: UserInfo?, event: AnalyticsBaseEvent? = null) {
        put(ADOBE_ATTR_APP_NAME, VALUE_GODTOOLS)
        put(ADOBE_ATTR_MARKETING_CLOUD_ID, adobeMarketingCloudId)

        put(ADOBE_ATTR_LOGGED_IN_STATUS, VALUE_NOT_LOGGED_IN)
        if (userProfile != null) {
            put(ADOBE_ATTR_LOGGED_IN_STATUS, VALUE_LOGGED_IN)
            put(ADOBE_ATTR_SSO_GUID, userProfile.ssoGuid)
            userProfile.grMasterPersonId?.let { put(ADOBE_ATTR_GR_MASTER_PERSON_ID, it) }
        }
        event?.run {
            locale?.let { put(ADOBE_ATTR_CONTENT_LANGUAGE, LocaleCompat.toLanguageTag(it)) }
            appSection?.let { put(ADOBE_ATTR_SITE_SECTION, it) }
            appSubSection?.let { put(ADOBE_ATTR_SITE_SUB_SECTION, it) }
        }
    }

    @WorkerThread
    private fun MutableMap<String, Any?>.stateContextData(userProfile: UserInfo?, event: AnalyticsScreenEvent) {
        baseContextData(userProfile, event)
        put(ADOBE_ATTR_SCREEN_NAME_PREVIOUS, previousScreenName)
        put(ADOBE_ATTR_SCREEN_NAME, event.screen)
    }

    // region Visitor ids
    private val userProfileStateFlow = oktaUserProfileProvider.userInfoFlow(refreshIfStale = false)
        .onEach { analyticsExecutor.execute { updateVisitorIdIdentifiers(it) } }
        .stateIn(coroutineScope, Eagerly, null)

    init {
        analyticsExecutor.execute {
            Visitor.syncIdentifier(
                VISITOR_ID_ECID,
                adobeMarketingCloudId,
                VisitorIDAuthenticationState.VISITOR_ID_AUTHENTICATION_STATE_UNKNOWN
            )
            updateVisitorIdIdentifiers(userProfileStateFlow.value)
        }
    }

    @WorkerThread
    private fun updateVisitorIdIdentifiers(userProfile: UserInfo?) {
        Visitor.syncIdentifiers(
            mutableMapOf(
                VISITOR_ID_GUID to userProfile?.ssoGuid,
                VISITOR_ID_MASTER_PERSON_ID to userProfile?.grMasterPersonId
            ), when {
                userProfile != null -> VisitorIDAuthenticationState.VISITOR_ID_AUTHENTICATION_STATE_AUTHENTICATED
                else -> VisitorIDAuthenticationState.VISITOR_ID_AUTHENTICATION_STATE_LOGGED_OUT
            }
        )
    }
    // endregion Visitor ids
}
